### PR TITLE
Factories upgrade

### PR DESCRIPTION
**Pull request by ntunin**
There was removed redundant functionality and improved naming. also implemented isEqual handler based on associated object

### CODE REVIEW

File: Example/CollectionViewToolsExample/Main/ContentViewState/TextViewState.swift [hunk 1]
```diff
@@ -2,11 +2,35 @@
...
+import Foundation
 
-    init(textContent: TextContent) {
-        self.textContent = textContent
-        super.init(content: textContent)
+final class TextViewState: ViewState, Expandable {
+    let text: String
+    let description: String
+    var isExpanded: Bool = false
+
```
Comment (Tublad) [inline Example/CollectionViewToolsExample/Main/ContentViewState/TextViewState.swift:11]: delete extra space

File: Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift [hunk 2]
```diff
@@ -5,29 +5,41 @@
 import CollectionViewTools
 
 protocol ContentViewCellItemFactoryOutput {
-    func reloadCollectionView()
-    func removeContentViewState(_ state: ContentViewState)
+    func updateContentView(for viewState: ViewState & Expandable)
+    func removeContentView(for viewState: ImageViewState)
```
Comment (Tublad) [inline Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift:9]:
```suggestion
    func updateContentViewState(_ viewState: ViewState & Expandable)
    func removeContentViewState(_ viewState: ImageViewState)
```
What do you think of this option?

File: Sources/Helpers/Factories/ComplexCellItemsFactory.swift [hunk 3]
```diff
@@ -0,0 +1,98 @@
+//
+//  ComplexCellItemFactory.swift
```
Comment (Tublad) [inline Sources/Helpers/Factories/ComplexCellItemsFactory.swift]: This is not superfluous here?

File: Sources/Helpers/Factories/ComplexCellItemsFactory.swift [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    /// - Parameters:
+    ///    - factory: a second cell item factory the associated type of which should be united
+    @discardableResult
+    public func factory(byJoining factory: ComplexCellItemsFactory) -> ComplexCellItemsFactory {
+        factory.factories.forEach { (key, value) in
+            factories[key] = value
+        }
+        return self
+    }
+
```
Comment (Tublad) [inline Sources/Helpers/Factories/ComplexCellItemsFactory.swift]: delete extra space

File: Sources/Helpers/Factories/ViewCellItemsFactory.swift [hunk 5]
```diff
@@ -4,34 +4,28 @@
...
     public typealias Cell = CollectionViewViewCell<View>
     public typealias CellItem = CollectionViewViewCellItem<Object, View>
-    typealias AssociatedFactory = AssociatedCellItemFactory<Object, Cell>
-
-    /// Set this handler to retrieve or create a specific set of cell items for the associated object
-    ///
-    /// - Parameters:
-    ///    - Object: the object associated with a cell item
-    public var initializationHandler: ((Object) -> [CollectionViewCellItem?])?
+    typealias Factory = CellItemsFactory<Object, Cell>
```
Comment (Tublad) [inline Sources/Helpers/Factories/ViewCellItemsFactory.swift:11]: Is there here a public not need?

Comment (ntunin) [reply Sources/Helpers/Factories/ViewCellItemsFactory.swift:11]: no, private moreover

Comment (Tublad) [reply Sources/Helpers/Factories/ViewCellItemsFactory.swift:11]: good

File: Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift (hunk 2 above)
Comment (antowkos) [reply Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift:9]: I think that initial methods naming seems more suitable

File: Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift [hunk 6]
```diff
@@ -61,7 +80,8 @@ final class ContentViewSectionItemsFactory {
 
         factory.viewConfigurationHandler = { view, _ in
             view.dividerHeight = 1
-            view.backgroundColor = .gray
+            view.dividerView.backgroundColor = .gray
+            view.dividerInsets = .init(top: 9, left: 0, bottom: 0, right: 0)
```
Comment (antowkos) [inline Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift:84]: Pretty strange top inset 🤔

File: Sources/Diff/DiffItem.swift [hunk 7]
```diff
@@ -2,6 +2,10 @@
 //  Copyright © 2019 Rosberry. All rights reserved.
 //
 
+import Foundation
+
+public typealias CanBeDiff = Equatable & CustomDebugStringConvertible
```
Comment (antowkos) [inline Sources/Diff/DiffItem.swift]: `DiffCompatible` seems more suitable

File: Sources/Helpers/Factories/CellItemFactory.swift [hunk 8]
```diff
@@ -4,36 +4,112 @@
...
+        let cellItem = ConcreteCellItem(object: object)
+        basicSetup(cellItem: cellItem, object: object)
+        return cellItem
+    }
+
+    /// Assigns cell configuration and size configuration handlers to factory handlers
+    ///
+    /// - Parameters:
+    ///    - object: an object to create a cell item for it
+    public func basicSetup<ConcreteCellItem: CellItem>(cellItem: ConcreteCellItem, object: Object) {
```
Comment (antowkos) [inline Sources/Helpers/Factories/CellItemFactory.swift]: Pretty strange name for a a function? Is there other any `setup` method?

File: Sources/Items/UniversalCollectionViewCellItem.swift [hunk 9]
```diff
@@ -46,9 +54,9 @@ public class UniversalCollectionViewCellItem<U: GenericDiffItem, T: UICollection
     }
 
     public func isEqual(to item: DiffItem) -> Bool {
-        guard let cellItem = item as? UniversalCollectionViewCellItem<U, T> else {
+        guard let cellItem = item as? UniversalCollectionViewCellItem<Object, Cell> else {
             return false
         }
-        return object.isEqual(to: cellItem)
+        return object == cellItem.object && object == originalObject && cellItem.object == originalObject
```
Comment (antowkos) [inline Sources/Items/UniversalCollectionViewCellItem.swift]: Please, align expressions on new line each

**Review by Antowkos — CHANGES_REQUESTED** [review]

File: Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift (hunk 2 above)
Comment (Tublad) [reply Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift:9]: ok

File: Sources/Items/UniversalCollectionViewCellItem.swift [hunk 10]
```diff
@@ -4,28 +4,36 @@
...
 
-    /// Provides an association of object that presents data model or view state of cell item
-    public let object: U
+    public let object: Object
+    private let originalObject: Object
 
-    required init(object: U) {
+    required init(object: Object) {
+        if let object = object as? NSCopying {
+            self.originalObject = object.copy(with: nil) as! Object
```
Comment (stason314) [inline Sources/Items/UniversalCollectionViewCellItem.swift]: Can you avoid force unwrap?

File: Sources/Helpers/Factories/ComplexCellItemsFactory.swift [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    public func unjoin<Object: CanBeDiff, Cell: UICollectionViewCell>(factory: CellItemsFactory<Object, Cell>) {
+        factories.removeValue(forKey: String(describing: Object.self))
+    }
+
+    /// Removes a factory from complex hierarchy
+    ///
+    /// - Parameters:
+    ///     - factory: a factory that should be removed
+    public func unjoin(factory: ComplexCellItemsFactory) {
+        factory.factories.forEach { (key, _) in
```
Comment (stason314) [inline Sources/Helpers/Factories/ComplexCellItemsFactory.swift]:
```suggestion
        factory.factories.forEach { key, _ in
```

**Review by stason314 — CHANGES_REQUESTED** [review]

File: Sources/Helpers/Factories/CellItemFactory.swift (hunk 8 above)
Comment (ntunin) [reply Sources/Helpers/Factories/CellItemFactory.swift]: could not remember a point of this, renamed to setup

File: Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift (hunk 6 above)
Comment (ntunin) [reply Example/CollectionViewToolsExample/Main/Factories/ContentViewSectionItemsFactory.swift:84]: To align divider in free space at center approximately

**Review by stason314 — APPROVED** [review]

**Review by Antowkos — APPROVED** [review]

**Pull request merged**